### DISCUSSION TITLE
Click performance is poor

### DISCUSSION THREAD

**Discussion by jonathan-s** [pallets/click#2692] [Q&A] [answered]
When I'm running click using a group I'm seeing pretty terrible performance. It's slow. 

```
my_command  0.74s user 0.24s system 29% cpu 3.307 total
```

The only thing this executes is the help output, so most of the code it should be executing would be click code. The output is `time my_command`. 

I don't know if I'm doing something wrong here myself, I'd be happy to take any suggestions for debugging the slowness here. 

Environment:

- Python version: 3.8.16
- Click version: 8.1.7

**Comment by ThiefMaster** [comment, marked as answer]
Are you sure your command isn't simply importing lots of stuff from your application? Deferring imports until you actually need them will likely give you a massive performance increase if you're importing a big application behind.

**Reply by jonathan-s** [reply]
Thanks for putting me on the right track!

**Reply by ThiefMaster** [reply]
You may find this useful, I use it in my own application for exactly that reason:

```py
class LazyGroup(click.Group):
    """
    A click Group that imports the actual implementation only when
    needed.  This allows for more resilient CLIs where the top-level
    command does not fail when a subcommand is broken enough to fail
    at import time.
    """

    def __init__(self, import_name, **kwargs):
        self._import_name = import_name
        super().__init__(**kwargs)

    @cached_property
    def _impl(self):
        module, name = self._import_name.split(':', 1)
        return getattr(import_module(module), name)

    def get_command(self, ctx, cmd_name):
        return self._impl.get_command(ctx, cmd_name)

    def list_commands(self, ctx):
        return self._impl.list_commands(ctx)

    def invoke(self, ctx):
        return self._impl.invoke(ctx)

    def get_usage(self, ctx):
        return self._impl.get_usage(ctx)

    def get_params(self, ctx):
        return self._impl.get_params(ctx)
```

An example of how I'm using it:

```py
@cli.group(cls=LazyGroup, import_name='indico.cli.setup:cli')
def setup():
    """Setup Indico."""
```

The `cli`  inside the referenced module is simply another click group.

**Reply by jonathan-s** [reply]
ah, that's pretty sweet! Thanks a lot!
